Implement load_into for SharedPtrDataLoader and add test (#11562)

Tanish2101 opened the pull request:
Summary:
Implemented the load_into method for SharedPtrDataLoader to support copying preallocated buffer data into a target buffer. This enables use cases where mutable model state has a meaningful initial value (e.g., on-device training or serialized read-before-write data).

Fixes
Fixes #11562

Test plan:
Added a unit test in shared_ptr_data_loader_test.cpp to verify that load_into():
1. Copies the correct data for valid offset/size.
2. Fails properly on invalid requests (e.g., out-of-bounds reads).

Tanish2101 committed:
Implement load_into for SharedPtrDataLoader and add test (#11562)

Tanish2101 commented:
@pytorchbot label "release notes: none"

Tanish2101 commented:
Hi @JacobSzwejbka, I created a PR for this issue. This is the first time I am contributing to executorch. Please Let me know if there are any changes required.

Tanish2101 commented:
I have implemented the load_into() function for **shared_ptr data loader** so the implementation done by @keyprocedure for  **mmap data loader** is sufficient? means there is no need for seperate implementation for shared_ptr ? 

@mergennachin

Tanish2101 committed:
Updated the shared_ptr_data_loader.h 

I have merge the declaration and implementation of 'load_into' function.

Tanish2101 committed:
Update shared_ptr_data_loader.h

JacobSzwejbka reviewed (approved):
Letting CI run, but looks good. Thanks for contributing!

JacobSzwejbka committed:
Merge branch 'main' into implement-sharedptr-loader

Tanish2101 committed:
Fix: Use valid Error enum in SharedPtrDataLoader and update test

- Replaced non-existent  with  in  implementation.
- Updated the corresponding unit test to match the corrected error.
- Ensures correctness and prevents enum resolution error during compilation.

Tanish2101 committed:
Merge branch 'pytorch:main' into implement-sharedptr-loader

Tanish2101 commented:
Fixed: Changed invalid error code Error::OutOfBounds to Error::InvalidArgument in both implementation and test. OutOfBounds is not part of the executorch::runtime::Error enum 

@JacobSzwejbka

lucylq commented:
Hi @Tanish2101 , are you still working on this? See that it's approved, do you want to rebase and merge?

mergennachin committed:
Merge branch 'main' into implement-sharedptr-loader
